src/aws-cpp-sdk-core/include/aws/core/client/AWSClientEventStreamingAsyncTask.h

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#pragma once
+
+#include <aws/core/Core_EXPORTS.h>
+#include <aws/core/NoResult.h>
+#include <aws/core/client/AWSClient.h>
+#include <aws/core/utils/event/EventDecoderStream.h>
+#include <aws/core/utils/event/EventEncoderStream.h>
+#include <aws/core/utils/event/EventStreamHandler.h>
+#include <aws/core/utils/logging/ErrorMacros.h>
+
+namespace Aws {
+namespace Client {
+
+class AsyncCallerContext;
+
+/**
+ * A template class to create bi-directional eventstreaming async task.
+ * This is essentially a C++14 capture-by-move lambda
+ */
+template <typename OutcomeT, typename ClientT, typename AWSEndpointT, typename RequestT, typename HandlerT>
+class AWS_CORE_LOCAL BidirectionalEventStreamingTask final {
+ public:
+  BidirectionalEventStreamingTask(const ClientT* pClientThis, AWSEndpointT&& endpoint, const std::shared_ptr<RequestT>& pRequest,
+                                  const HandlerT& handler, const std::shared_ptr<const Aws::Client::AsyncCallerContext>& handlerContext,
+                                  const std::shared_ptr<Utils::Event::EventEncoderStream>& stream, const Http::HttpMethod method,
+                                  char const* const signerName)
+      : m_clientThis(pClientThis),
+        m_endpoint(std::move(endpoint)),
+        m_pRequest(pRequest),
+        m_handler(handler),
+        m_handlerContext(handlerContext),
+        m_stream(stream),
+        m_method(method),
+        m_signerName(signerName),
+        m_sem(Aws::MakeShared<Aws::Utils::Threading::Semaphore>("BidirectionalEventStreamingTask", 0, 1)) {
```

Comment:
why do we want to use a semaphore and not a mutex if max value of semaphore is 1?

Reply:
That's the original generated design code.
We could switch to a singal, but it will take 2 more member variables (mutex and a conditional_variable) to pass around instead of this single one.